src/OpenTelemetry.Api/Context/BaggageContext.cs

```diff
@@ -0,0 +1,241 @@
+        /// </summary>
+        /// <param name="name">Baggage item name.</param>
+        /// <returns>Baggage item or <see langword="null"/> if nothing was found.</returns>
+        public object GetBaggage(string name)
+        {
+            return this.baggage.TryGetValue(name, out string value)
+                ? value
+                : null;
+        }
+
+        /// <summary>
+        /// Returns a new <see cref="BaggageContext"/> which contains the new key/value pair.
+        /// </summary>
+        /// <param name="name">Baggage item name.</param>
+        /// <param name="value">Baggage item value.</param>
+        /// <returns>New <see cref="BaggageContext"/> containing the key/value pair.</returns>
+        public BaggageContext SetBaggage(string name, string value)
+        {
+            var baggageContext = new BaggageContext(
+                new Dictionary<string, string>(this.baggage, StringComparer.OrdinalIgnoreCase)
+                {
+                    [name] = value,
+                });
+
+            Current = baggageContext;
+
+            return baggageContext;
+        }
+
+        /// <summary>
+        /// Returns a new <see cref="BaggageContext"/> which contains the new key/value pair.
+        /// </summary>
+        /// <param name="baggage">Baggage key/value pairs.</param>
+        /// <returns>New <see cref="BaggageContext"/> containing the key/value pair.</returns>
+        public BaggageContext SetBaggage(params KeyValuePair<string, string>[] baggage)
+            => this.SetBaggage((IEnumerable<KeyValuePair<string, string>>)baggage);
+
+        /// <summary>
+        /// Returns a new <see cref="BaggageContext"/> which contains the new key/value pair.
+        /// </summary>
+        /// <param name="baggage">Baggage key/value pairs.</param>
+        /// <returns>New <see cref="BaggageContext"/> containing the key/value pair.</returns>
+        public BaggageContext SetBaggage(IEnumerable<KeyValuePair<string, string>> baggage)
+        {
+            var newBaggage = new Dictionary<string, string>(this.baggage, StringComparer.OrdinalIgnoreCase);
+
+            foreach (var item in baggage)
+            {
+                newBaggage[item.Key] = item.Value;
+            }
+
+            var baggageContext = new BaggageContext(newBaggage);
+            Current = baggageContext;
+            return baggageContext;
+        }
+
+        /// <summary>
+        /// Returns a new <see cref="BaggageContext"/> with the key/value pair removed.
+        /// </summary>
+        /// <param name="name">Baggage item name.</param>
+        /// <returns>New <see cref="BaggageContext"/> containing the key/value pair.</returns>
+        public BaggageContext RemoveBaggage(string name)
+        {
+            var baggage = new Dictionary<string, string>(this.baggage, StringComparer.OrdinalIgnoreCase);
+            baggage.Remove(name);
```

Comment:
We might want to consider HAMT https://www.python.org/dev/peps/pep-0550/#implementation for perf, that can be a separate perf improvement PR.

Reply:
Ya HAMT 🐹 didn't jump into my brain but I was definitely thinking as I went through this that the Dictionary probably isn't the best data structure. Lots of copying going on. After we get the API nailed down I'll do some perf hacking in a follow-up PR.